### PR TITLE
Break-up unit tests under `SampleAppSpec` into feature-grouped specs #18

### PR DESCRIPTION
**Pull request by connected-akhan**
- refactored tests by feature into separate classes

### CODE REVIEW

File: UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift [hunk 1]
```diff
@@ -0,0 +1,60 @@
+//
+//  VCFromNibSpec.swift
+//  UnIt_SampleAppTests
+//
+//  Created by Amaan Khan on 2019-07-18.
+//  Copyright © 2019 cl-dev. All rights reserved.
+//
```
Comment (connected-swu) [inline UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift]: Hmm, I think we should drop these headers

File: UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+import Foundation
+import Nimble
+import Quick
+import UnIt
+@testable import UnIt_SampleApp
+
+class VCFromNibSpec: QuickSpec {
```
Comment (connected-swu) [inline UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift]: Since a `UIViewController` extension to from nib is being tested here, I think the name of this spec should be something like `UIViewController+LoadingSpec` instead.

File: UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    override func spec() {
+        var nibSubject : NibViewController!
```
Comment (connected-swu) [inline UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift]: Just `subject`?

File: UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            }
+            
+            it("should not be nil") {
+                expect(nibSubject).notTo(beNil())
+                expect(nibSubject.view).notTo(beNil())
+            }
+            
+            it("should have a label that has text: 'Top View Controller Label'") {
+                expect(nibSubject.view.firstLabel(with: "Top View Controller Label")).notTo(beNil())
+
```
Comment (connected-swu) [inline UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift]: extraneous blank space =P  being super nitpicky haha

File: UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            
+            it("should have a label with the text 'Left Stack Label' in a UIStackView") {
+                expect(nibSubject.view.firstLabel(with: "Left Stack Label")).notTo(beNil())
+            }
+            
+            it("should have a button with the text 'Right Stack Button' in a UIStackView") {
+                expect(nibSubject.view.firstButton(with: "Right Stack Button")).notTo(beNil())
+            }
+        }
+    }
```
Comment (connected-swu) [inline UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift]: What are your thoughts about the comprehensiveness of testing on this API's here?

File: UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift (hunk 2 above)
Comment (connected-akhan) [reply UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift]: Yea that's good. Also, there is loading from nib vs. storyboard. To distinguish between the two, would `UIViewController+LoadingFromNibSpec` and `UIViewController+LoadingFromStoryboardSpec` be a good idea?

File: UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift (hunk 3 above)
Comment (connected-akhan) [reply UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift]: or `sut`?

File: UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift (hunk 2 above)
Comment (connected-swu) [reply UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift]: I think that might be too granular – they seem to belong in the same feature group and the functionality hasn't grown out enough to warrant splitting into 2 specs in my mind.  Will let you decide.

File: UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift (hunk 3 above)
Comment (connected-swu) [reply UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift]: I usually go with `subject`, but `sut` also commonly used – you pick.  Whichever we pick, just make sure it's consistent across the board.

File: UnIt_SampleApp/UnIt_SampleAppTests/VCFromStoryboardSpec.swift [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        
+        describe("ViewController from storyboard") {
+            beforeEach() {
+                subject = UIViewController.loadAndSetupViewControllerFromStoryboard("Main", "ViewController", Device.iPhoneXSMax)
+            }
+            
+            it("should not be nil") {
+                expect(subject).notTo(beNil())
+                expect(subject.view).notTo(beNil())
+            }
```
Comment (connected-swu) [inline UnIt_SampleApp/UnIt_SampleAppTests/VCFromStoryboardSpec.swift]: Oh I see how you're splitting this up.

Instead of splitting into **use-case groups** of where the view-controller is instantiated from, we should split by **feature groups** of API feature types.

File: UnIt_SampleApp/UnIt_SampleAppTests/VCWithConstraintBreaksSpec.swift [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+//  Copyright © 2019 cl-dev. All rights reserved.
+//
+
+import Foundation
+import Nimble
+import Quick
+import UnIt
+@testable import UnIt_SampleApp
+
+class VCWithConstraintBreaksSpec: QuickSpec {
```
Comment (connected-swu) [inline UnIt_SampleApp/UnIt_SampleAppTests/VCWithConstraintBreaksSpec.swift]: This is an extension of `UIViewController`, think about renaming this to something like `UIViewController+ConstraintBreakDetectingSpec.swift`

File: UnIt_SampleApp/UnIt_SampleAppTests/VCWithLabelTruncationSpec.swift [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+//  Copyright © 2019 cl-dev. All rights reserved.
+//
+
+import Foundation
+import Nimble
+import Quick
+import UnIt
+@testable import UnIt_SampleApp
+
+class VCWithLabelTruncationSpec: QuickSpec {
```
Comment (connected-swu) [inline UnIt_SampleApp/UnIt_SampleAppTests/VCWithLabelTruncationSpec.swift]: I think naming can be improved overall.

File: UnIt_SampleApp/UnIt_SampleAppTests/VCWithLabelTruncationSpec.swift [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+                    expect(labelSubject.threeLineLabelWithTruncation.isTruncated).to(equal(true))
+                }
+                
+                it("should determine the theoretical number of lines the label should have to look untruncated") {
+                    expect(labelSubject.oneLineLabelNoTruncation.numberOfTheoreticalLines()).to(equal(1))
+                    expect(labelSubject.oneLineLabelWithTruncation.numberOfTheoreticalLines()).to(equal(2))
+                    expect(labelSubject.twoLineLabelNoTruncation.numberOfTheoreticalLines()).to(equal(2))
+                    expect(labelSubject.oneLineLabelWithIncreasingFontToTruncate.numberOfTheoreticalLines()).to(equal(3))
+                    expect(labelSubject.oneLineLabelThatSizeWillDecreaseToTruncate.numberOfTheoreticalLines()).to(equal(2))
+                    expect(labelSubject.threeLineLabelWithTruncation.numberOfTheoreticalLines()).to(equal(5))
```
Comment (connected-swu) [inline UnIt_SampleApp/UnIt_SampleAppTests/VCWithLabelTruncationSpec.swift]: Wonder if this should be broken down into a context, hmmm

File: UnIt_SampleApp/UnIt_SampleAppTests/VCWithLabelTruncationSpec.swift [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+            context("when the labels are set") {
+                it("should determine the theoretical number of lines the label should have to look untruncated") {
+                    expect(labelSubject.threeLineLabelWithTruncation.numberOfTheoreticalLines()).to(equal(4)) // same as line 227
+                }
+            }
+        }
+        
+        describe("View Controller with untruncated and truncated labels on the smallest possible iPhone to test screen sizes") {
+            beforeEach {
+                labelSubject = UIViewController.loadAndSetupViewControllerFromNib("LabelViewController", LabelViewController.self, Device.iPhoneSE)
```
Comment (connected-swu) [inline UnIt_SampleApp/UnIt_SampleAppTests/VCWithLabelTruncationSpec.swift]: We're probably missing test coverage on screen sizes.

File: UnIt_SampleApp/UnIt_SampleAppTests/VCWithOffscreenViewsSpec.swift [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+                offscreenSubject = UIViewController.loadAndSetupViewControllerFromNib("OffScreenViewController", OffScreenViewController.self, Device.iPhone7)
+            }
+            
+            context("When there are offscreen elements") {
+                it("should find all the offscreen elements") {
+                    expect(offscreenSubject.viewsOffScreen().count).to(equal(6))
+                }
+                
+                it("should find all the partially offscreen elements") {
+                    expect(offscreenSubject.viewsPartiallyOffScreen().count).to(equal(5))
```
Comment (connected-swu) [inline UnIt_SampleApp/UnIt_SampleAppTests/VCWithOffscreenViewsSpec.swift]: Interesting that we have 5, should dig in later to figure out what the boundary cases are here.

File: UnIt_SampleApp/UnIt_SampleAppTests/VCWithOffscreenViewsSpec.swift [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+        var offscreenSubject: OffScreenViewController!
+        
+        describe("View Controller with views offscreen") {
+            beforeEach {
+                offscreenSubject = UIViewController.loadAndSetupViewControllerFromNib("OffScreenViewController", OffScreenViewController.self, Device.iPhone7)
+            }
+            
+            context("When there are offscreen elements") {
+                it("should find all the offscreen elements") {
+                    expect(offscreenSubject.viewsOffScreen().count).to(equal(6))
```
Comment (connected-swu) [inline UnIt_SampleApp/UnIt_SampleAppTests/VCWithOffscreenViewsSpec.swift]: 6 implies to me there might be edge cases contained within the view controller designed to test this API, we might want to be more specific about these tests or break them up into individual parts if possible.

File: UnIt_SampleApp/UnIt_SampleAppTests/VCWithOverlappingElementsSpec.swift [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+                }
+                
+                it("should find all the sibiling overlapping views in the view controller that are in the view tag white list") {
+                    let whiteListedOverlapViews = overlapSubject.overlappingSubviews(whiteList: [11,12])
+                    expect(whiteListedOverlapViews.count).to(equal(3))
+                }
+            }
+        }
+    }
+}
```
Comment (connected-swu) [inline UnIt_SampleApp/UnIt_SampleAppTests/VCWithOverlappingElementsSpec.swift]: Yea, overall the wording and naming of tests can be made much better.

File: UnIt_SampleApp/UnIt_SampleAppTests/VCWithUserInputsSpec.swift [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+                beforeEach {
+                    actionSubject.numericTextField.paste(with: "1a2b3c")
+                    actionSubject.numericTextField.paste(with: "12345")
+                }
+                
+                it("should display only '12345' in the numeric text field") {
+                    expect(actionSubject.numericTextField.text).to(be("12345"))
+                }
+            }
+        }
```
Comment (connected-swu) [inline UnIt_SampleApp/UnIt_SampleAppTests/VCWithUserInputsSpec.swift]: I think coverage could be improved here

File: UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift (hunk 3 above)
Comment (connected-akhan) [reply UnIt_SampleApp/UnIt_SampleAppTests/VCFromNibSpec.swift]: `subject` it is!

File: UnIt_SampleApp/UnIt_SampleAppTests/VCWithOffscreenViewsSpec.swift [hunk 15]
```diff
@@ -0,0 +1,12 @@
+import Foundation
+import Nimble
+import Quick
+import UnIt
+@testable import UnIt_SampleApp
+
+class VCWithOffscreenViewsSpec: QuickSpec {
```
Comment (connected-swu) [inline UnIt_SampleApp/UnIt_SampleAppTests/VCWithOffscreenViewsSpec.swift]: Is this spec still needed?  Not yet renamed?

File: UnIt_SampleApp/UnIt_SampleAppTests/UIView+InspectingSpec.swift [hunk 16]
```diff
@@ -0,0 +1,24 @@
+import Foundation
```
Comment (connected-swu) [inline UnIt_SampleApp/UnIt_SampleAppTests/UIView+InspectingSpec.swift]: Hmmm, we did a sweep to make sure our imports are consistent yes?

File: UnIt/Source/UIViewController+Loading.swift [hunk 17]
```diff
@@ -0,0 +1,28 @@
+import Foundation
+import UIKit
```
Comment (connected-swu) [inline UnIt/Source/UIViewController+Loading.swift]: consistency in imports?

File: UnIt_SampleApp/UnIt_SampleAppTests/UIView+FindingBasedOnTextSpec.swift [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+            // what about table view, list view?
+            
+            it("should have a label with the text 'Left Stack Label' in a UIStackView") {
+                expect(vcFromNib.view.firstLabel(with: "Left Stack Label")).notTo(beNil())
+            }
+            
+            it("should have a button with the text 'Right Stack Button' in a UIStackView") {
+                expect(vcFromNib.view.firstButton(with: "Right Stack Button")).notTo(beNil())
+            }
+        }
```
Comment (connected-swu) [inline UnIt_SampleApp/UnIt_SampleAppTests/UIView+FindingBasedOnTextSpec.swift]: - Does the naming of `vcFromNib` and `vcFromStoryboard` best convey test intent? 
- Also, why double up the tests between against storyboard-vc and nib-vc?

File: UnIt_SampleApp/UnIt_SampleAppTests/UIViewController+InspectingSpec.swift [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+                }
+            }
+        }
+        
+        describe("View Controller with overlapping elements") {
+            beforeEach {
+                vcWithOverlappingElements = UIViewController.loadAndSetupViewControllerFromNib("OverlapViewController", OverlapViewController.self, Device.iPhone6)
+            }
+            
+            // TODO-AK
```
Comment (connected-swu) [inline UnIt_SampleApp/UnIt_SampleAppTests/UIViewController+InspectingSpec.swift]: ?

**Comment by connected-swu** [comment]
Good to merge after minor changes in comments.

File: UnIt_SampleApp/UnIt_SampleAppTests/UIView+InspectingSpec.swift (hunk 16 above)
Comment (connected-akhan) [reply UnIt_SampleApp/UnIt_SampleAppTests/UIView+InspectingSpec.swift]: I tried to apply the rule of import only what is needed, so now, most files import neither Foundation nor UIKit, and it still compiles and all tests pass. So now it is consistent across all files as well.

File: UnIt_SampleApp/UnIt_SampleAppTests/UIView+FindingBasedOnTextSpec.swift (hunk 18 above)
Comment (connected-akhan) [reply UnIt_SampleApp/UnIt_SampleAppTests/UIView+FindingBasedOnTextSpec.swift]: Thanks, you're right, since we are testing the API, and not the different types of VC, it is sufficient to have tests for just one VC. Removed duplicate tests.

File: UnIt_SampleApp/UnIt_SampleAppTests/VCWithOffscreenViewsSpec.swift (hunk 15 above)
Comment (connected-akhan) [reply UnIt_SampleApp/UnIt_SampleAppTests/VCWithOffscreenViewsSpec.swift]: Ah, when I was trying to delete it, I must have clicked remove reference instead of move to trash, so it went away in Xcode and remained in the actual folder. Deleted file from folder.
